File: android_mobile/src/main/java/com/alexstyl/specialdates/dailyreminder/DailyReminderNotifier.java
package com.alexstyl.specialdates.dailyreminder;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.support.v4.app.NotificationCompat;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.StyleSpan;

import com.alexstyl.android.Version;
import com.alexstyl.resources.ColorResources;
import com.alexstyl.resources.DimensionResources;
import com.alexstyl.specialdates.Strings;
import com.alexstyl.specialdates.Optional;
import com.alexstyl.specialdates.R;
import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.date.ContactEvent;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.events.bankholidays.BankHoliday;
import com.alexstyl.specialdates.events.namedays.activity.NamedayActivity;
import com.alexstyl.specialdates.images.ImageLoader;
import com.alexstyl.specialdates.upcoming.UpcomingEventsActivity;
import com.alexstyl.specialdates.util.NaturalLanguageUtils;

import java.util.ArrayList;
import java.util.List;

public final class DailyReminderNotifier {

    private static final int NOTIFICATION_ID_DAILY_REMINDER_CONTACTS = 0;
    private static final int NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS = 1;
    private static final int NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS = 2;
    private static final int MAX_CONTACTS = 3;

    private final Context context;
    private final NotificationManager notificationManager;
    private final ImageLoader imageLoader;
    private final Strings strings;
    private final DimensionResources dimensions;
    private final ColorResources colorResources;
    private final DailyReminderPreferences preferences;

    DailyReminderNotifier(Context context,
                          NotificationManager notificationManager,
                          ImageLoader imageLoader,
                          Strings strings,
                          ColorResources colorResources,
                          DimensionResources dimensions,
                          DailyReminderPreferences preferences) {
        this.notificationManager = notificationManager;
        this.strings = strings;
        this.imageLoader = imageLoader;
        this.context = context.getApplicationContext();
        this.dimensions = dimensions;
        this.colorResources = colorResources;
        this.preferences = preferences;
    }

    void forDailyReminder(Date date, List<ContactEvent> events) {
        Bitmap largeIcon = null;
        int contactCount = events.size();

        if (shouldDisplayContactImage(contactCount)) {
            Contact displayingContact = events.get(0).getContact();
            int size = dimensions.getPixelSize(android.R.dimen.notification_large_icon_width);
            Optional<Bitmap> loadedIcon =
                    imageLoader.load(displayingContact.getImagePath())
                            .withSize(size, size)
                            .synchronously();
            if (Version.hasLollipop() && loadedIcon.isPresent()) {
                // in Lollipop the notifications is the default to use Rounded Images
                largeIcon = getCircleBitmap(loadedIcon.get());
            }
        }

        Intent startIntent = UpcomingEventsActivity.getStartIntent(context);
        PendingIntent intent =
                PendingIntent.getActivity(
                        context, NOTIFICATION_ID_DAILY_REMINDER_CONTACTS,
                        startIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );

        List<Contact> contacts = new ArrayList<>();
        for (ContactEvent event : events) {
            contacts.add(event.getContact());
        }
        String title = NaturalLanguageUtils.joinContacts(strings, contacts, MAX_CONTACTS);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context)
                .setSmallIcon(R.drawable.ic_stat_memento)
                .setContentTitle(title)
                .setLargeIcon(largeIcon)
                .setVisibility(NotificationCompat.VISIBILITY_PRIVATE)
                .setAutoCancel(true)
                .setContentIntent(intent)
                .setNumber(events.size())
                .setColor(colorResources.getColor(R.color.main_red))
                .setSound(preferences.getRingtoneSelected());

        if (events.size() == 1) {
            ContactEvent event = events.get(0);
            String msg = event.getLabel(date, strings);

            NotificationCompat.BigTextStyle bigTextStyle = new NotificationCompat.BigTextStyle().bigText(msg);
            bigTextStyle.setBigContentTitle(title);
            builder.setContentText(msg);

            builder.setStyle(bigTextStyle);

        } else if (contacts.size() > 1) {
            NotificationCompat.InboxStyle inboxStyle = new NotificationCompat.InboxStyle();
            inboxStyle.setBigContentTitle(title);

            for (int i = 0; i < events.size(); ++i) {

                ContactEvent event = events.get(i);
                Contact contact = event.getContact();
                String name = contact.getDisplayName().toString();

                String lineFormatted = name + "\t\t" + event.getLabel(date, strings);

                Spannable sb = new SpannableString(lineFormatted);
                sb.setSpan(new StyleSpan(Typeface.BOLD), 0, name.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                inboxStyle.addLine(sb);
            }

            builder.setStyle(inboxStyle);
            builder.setContentText(TextUtils.join(", ", contacts));
        }

        if (supportsPublicNotifications()) {
            String publicTitle = context.getString(R.string.contact_celebration_count, contactCount);
            NotificationCompat.Builder publicNotification = new NotificationCompat.Builder(context)
                    .setSmallIcon(R.drawable.ic_stat_memento)
                    .setAutoCancel(true)
                    .setContentIntent(intent)
                    .setContentTitle(publicTitle)
                    .setColor(colorResources.getColor(R.color.main_red));

            builder.setPublicVersion(publicNotification.build());
        }

        if (preferences.getVibrationSet()) {
            builder.setDefaults(Notification.DEFAULT_VIBRATE);
        }

        Notification notification = builder.build();

        notificationManager.notify(NOTIFICATION_ID_DAILY_REMINDER_CONTACTS, notification);

    }

    private static Bitmap getCircleBitmap(Bitmap bitmap) {
        final Bitmap output = Bitmap.createBitmap(
                bitmap.getWidth(),
                bitmap.getHeight(),
                Bitmap.Config.ARGB_8888
        );
        final Canvas canvas = new Canvas(output);

        final int color = Color.RED;
        final Paint paint = new Paint();
        final Rect rect = new Rect(0, 0, bitmap.getWidth(), bitmap.getHeight());
        final RectF rectF = new RectF(rect);

        paint.setAntiAlias(true);
        canvas.drawARGB(0, 0, 0, 0);
        paint.setColor(color);
        canvas.drawOval(rectF, paint);

        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawBitmap(bitmap, rect, rect, paint);

        return output;
    }

    private boolean supportsPublicNotifications() {
        return Version.hasLollipop();
    }

    private boolean shouldDisplayContactImage(int contactCount) {
        return contactCount == 1;
    }

    void forNamedays(List<String> names, Date date) {
        if (names == null || names.isEmpty()) {
            return;
        }
        PendingIntent intent = PendingIntent.getActivity(
                context, NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS,
                NamedayActivity.Companion.getStartIntent(context, date),
                PendingIntent.FLAG_UPDATE_CURRENT
        );

        String subtitle = NaturalLanguageUtils.join(strings, names, 1);
        String fullsubtitle = TextUtils.join(", ", names);

        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(context)
                .setSmallIcon(R.drawable.ic_stat_namedays)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(fullsubtitle))
                .setContentTitle(strings.todaysNamedays(names.size()))
                .setContentText(subtitle)
                .setAutoCancel(true)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setColor(colorResources.getColor(R.color.nameday_blue))
                .setContentIntent(intent);
        if (names.size() > 1) {
            mBuilder.setNumber(names.size());
        }
        notificationManager.notify(NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS, mBuilder.build());
    }

    public void cancelAllEvents() {
        notificationManager.cancel(NOTIFICATION_ID_DAILY_REMINDER_CONTACTS);
        notificationManager.cancel(NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS);
        notificationManager.cancel(NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS);

    }

    void forBankholiday(BankHoliday bankHoliday) {
        PendingIntent intent = PendingIntent.getActivity(
                context, NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS,
                UpcomingEventsActivity.getStartIntent(context),
                PendingIntent.FLAG_UPDATE_CURRENT
        );

        String subtitle = bankHoliday.getHolidayName();
        CharSequence title = context.getString(R.string.dailyreminder_title_bankholiday);
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context)
                .setSmallIcon(R.drawable.ic_stat_bankholidays)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(subtitle))
                .setContentTitle(title)
                .setContentText(subtitle)
                .setAutoCancel(true)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setColor(colorResources.getColor(R.color.bankholiday_green))
                .setContentIntent(intent);
        notificationManager.notify(NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS, builder.build());
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/dailyreminder/DailyReminderNotifier.kt
package com.alexstyl.specialdates.dailyreminder

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.graphics.RectF
import android.support.v4.app.NotificationCompat
import android.text.Spannable
import android.text.SpannableString
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import com.alexstyl.android.Version
import com.alexstyl.resources.Colors
import com.alexstyl.resources.DimensionResources
import com.alexstyl.specialdates.BuildConfig
import com.alexstyl.specialdates.R
import com.alexstyl.specialdates.Strings
import com.alexstyl.specialdates.contact.Contact
import com.alexstyl.specialdates.date.ContactEvent
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.events.bankholidays.BankHoliday
import com.alexstyl.specialdates.events.namedays.activity.NamedayActivity
import com.alexstyl.specialdates.home.HomeActivity
import com.alexstyl.specialdates.images.ImageLoader
import com.alexstyl.specialdates.person.PersonActivity
import com.alexstyl.specialdates.util.NaturalLanguageUtils
import java.util.ArrayList

class DailyReminderNotifier constructor(private val context: Context,
                                        private val notificationManager: NotificationManager,
                                        private val imageLoader: ImageLoader,
                                        private val strings: Strings,
                                        private val colors: Colors,
                                        private val dimensions: DimensionResources,
                                        private val preferences: DailyReminderPreferences) {


    fun forDailyReminder(date: Date, events: List<ContactEvent>) {
        createDailyReminderChannel()
        val width = context.resources.getDimensionPixelSize(android.R.dimen.notification_large_icon_width)
        val height = context.resources.getDimensionPixelSize(android.R.dimen.notification_large_icon_height)

        val contacts = ArrayList<Contact>()
        for ((_, _, _, contact) in events) {
            contacts.add(contact)
        }

        events.forEach { event ->
            val startIntent = PersonActivity.buildIntentFor(context, event.contact)
            val pendingIntent = PendingIntent.getActivity(
                    context,
                    NOTIFICATION_ID_DAILY_REMINDER_CONTACTS + event.contact.contactID.toInt(),
                    startIntent,
                    PendingIntent.FLAG_UPDATE_CURRENT)

            notificationManager.notify(
                    NOTIFICATION_ID_DAILY_REMINDER_CONTACTS + event.contact.contactID.toInt(),
                    NotificationCompat.Builder(context, CHANNEL_DAY_REMINDER)
                            .setSmallIcon(R.drawable.ic_stat_memento)
                            .setContentTitle(event.contact.displayName.toString())
                            .setContentText(createTextFor(event, date))
                            .setContentIntent(pendingIntent)
                            .setColor(colors.getDailyReminderColor())
                            .setSound(preferences.ringtoneSelected)
                            .setGroup(NOTIFICATION_GROUP_DAILY_REMINDER)
                            .apply {
                                imageLoader
                                        .load(event.contact.imagePath)
                                        .withSize(width, height)
                                        .synchronously().apply {
                                            if (isPresent) {
                                                setLargeIcon(get().toCircle())
                                            }
                                        }
                            }.apply {
                                if (supportsPublicNotifications()) {
                                    val publicTitle = context.getString(R.string.contact_celebration_count, contacts.size)
                                    setPublicVersion(
                                            NotificationCompat.Builder(context, CHANNEL_DAY_REMINDER)
                                                    .setSmallIcon(R.drawable.ic_stat_memento)
                                                    .setAutoCancel(true)
                                                    .setContentIntent(pendingIntent)
                                                    .setContentTitle(publicTitle)
                                                    .setColor(colors.getDailyReminderColor())
                                                    .build())
                                }
                            }
                            .build()
            )
        }
        val startIntent = HomeActivity.getStartIntent(context)
        val pendingIntent = PendingIntent.getActivity(
                context,
                NOTIFICATION_ID_DAILY_REMINDER_CONTACTS,
                startIntent,
                PendingIntent.FLAG_UPDATE_CURRENT)


        val groupBuilder = NotificationCompat.Builder(context, CHANNEL_DAY_REMINDER)
                .setContentTitle(NaturalLanguageUtils.joinContacts(strings, contacts, MAX_CONTACTS))
                // .setContentText("Content Text")
                .setGroupSummary(true)
                .setGroup(NOTIFICATION_GROUP_DAILY_REMINDER)
                .setColor(colors.getDailyReminderColor())
                .setSmallIcon(R.drawable.ic_stat_memento)
                .setContentIntent(pendingIntent)

        notificationManager.notify(NOTIFICATION_ID_DAILY_REMINDER_CONTACTS, groupBuilder.build())
    }

    private fun createTextFor(it: ContactEvent, date: Date): SpannableString {
        val label = it.getLabel(date, strings)
        return SpannableString(label).apply {
            setSpan(ForegroundColorSpan(colors.getColorFor(it.type)), 0, label.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        }
    }

    private fun createDailyReminderChannel() {
        if (Version.hasOreo()) {
            notificationManager.createNotificationChannel(
                    NotificationChannel(
                            DailyReminderNotifier.CHANNEL_DAY_REMINDER,
                            strings.dailyReminder(),
                            NotificationManager.IMPORTANCE_DEFAULT
                    )
            )
        }
    }

    private fun supportsPublicNotifications(): Boolean {
        return Version.hasLollipop()
    }

    fun forNamedays(names: List<String>, date: Date) {
        if (names.isEmpty()) {
            return
        }
        val intent = PendingIntent.getActivity(
                context, NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS,
                NamedayActivity.getStartIntent(context, date),
                PendingIntent.FLAG_UPDATE_CURRENT
        )

        val subtitle = NaturalLanguageUtils.join(strings, names, 1)
        val fullsubtitle = TextUtils.join(", ", names)

        val builder = NotificationCompat.Builder(context, CHANNEL_NAMEDAYS)
                .setSmallIcon(R.drawable.ic_stat_namedays)
                .setStyle(NotificationCompat.BigTextStyle().bigText(fullsubtitle))
                .setContentTitle(strings.todaysNamedays(names.size))
                .setContentText(subtitle)
                .setAutoCancel(true)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setColor(colors.getNamedaysColor())
                .setContentIntent(intent)
        if (names.size > 1) {
            builder.setNumber(names.size)
        }
        createNamedaysChannel()
        notificationManager.notify(NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS, builder.build())
    }

    private fun createNamedaysChannel() {
        if (Version.hasOreo()) {
            notificationManager.createNotificationChannel(
                    NotificationChannel(
                            DailyReminderNotifier.CHANNEL_NAMEDAYS,
                            strings.namedays(),
                            NotificationManager.IMPORTANCE_DEFAULT
                    )
            )
        }
    }

    fun cancelAllEvents() {
        notificationManager.cancel(NOTIFICATION_ID_DAILY_REMINDER_CONTACTS)
        notificationManager.cancel(NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS)
        notificationManager.cancel(NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS)

    }

    fun forBankholiday(bankHoliday: BankHoliday) {
        val intent = PendingIntent.getActivity(
                context, NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS,
                HomeActivity.getStartIntent(context),
                PendingIntent.FLAG_UPDATE_CURRENT
        )

        val subtitle = bankHoliday.holidayName
        val title = context.getString(R.string.dailyreminder_title_bankholiday)
        val builder = NotificationCompat.Builder(context, CHANNEL_BANK_HOLIDAYS)
                .setSmallIcon(R.drawable.ic_stat_bankholidays)
                .setStyle(NotificationCompat.BigTextStyle().bigText(subtitle))
                .setContentTitle(title)
                .setContentText(subtitle)
                .setAutoCancel(true)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setColor(colors.getBankholidaysColor())
                .setContentIntent(intent)

        createBankholidayChannel()
        notificationManager.notify(NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS, builder.build())
    }

    private fun createBankholidayChannel() {
        if (Version.hasOreo()) {
            notificationManager.createNotificationChannel(
                    NotificationChannel(
                            DailyReminderNotifier.CHANNEL_BANK_HOLIDAYS,
                            strings.bankholidays(),
                            NotificationManager.IMPORTANCE_DEFAULT
                    )
            )
        }
    }

    companion object {

        private const val NOTIFICATION_ID_DAILY_REMINDER_CONTACTS = 0
        private const val NOTIFICATION_ID_DAILY_REMINDER_NAMEDAYS = 1
        private const val NOTIFICATION_ID_DAILY_REMINDER_BANKHOLIDAYS = 2
        private const val NOTIFICATION_GROUP_DAILY_REMINDER = "reminder_group"
        private const val MAX_CONTACTS = 3
        const val CHANNEL_DAY_REMINDER = BuildConfig.APPLICATION_ID + ".channel:daily_reminder"
        const val CHANNEL_NAMEDAYS = BuildConfig.APPLICATION_ID + ".channel:namedays"
        const val CHANNEL_BANK_HOLIDAYS = BuildConfig.APPLICATION_ID + ".channel:bank_holidays"

        private fun getCircleBitmap(bitmap: Bitmap): Bitmap {
            val output = Bitmap.createBitmap(
                    bitmap.width,
                    bitmap.height,
                    Bitmap.Config.ARGB_8888
            )
            val canvas = Canvas(output)

            val color = Color.RED
            val paint = Paint()
            val rect = Rect(0, 0, bitmap.width, bitmap.height)
            val rectF = RectF(rect)

            paint.isAntiAlias = true
            canvas.drawARGB(0, 0, 0, 0)
            paint.color = color
            canvas.drawOval(rectF, paint)

            paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
            canvas.drawBitmap(bitmap, rect, rect, paint)

            return output
        }
    }

    private fun Bitmap.toCircle(): Bitmap? = getCircleBitmap(this)
}

